finished validation issue number:240

guodong323 opened the pull request.

ddfridley reviewed (changes requested):
First, please install [prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) in visual studio code, so that when you save the file it will be formatted to the style set for this project.  It makes it easier to review changes.

Then, if you `npm run test` there will be many test failures.  And you can `npm run test dturn` and it will run fewer tests and that might help.

We need to make all tests pass.

I looked into this one: npm run test app/dturn/__tests__/rank-statements.js
The problem in that insertStatementIds is failing with
```
  console.error                                                                                                                                  
    User user0 is not part of discussion testRankingDiscussion   
```
For insertStatementId we need to allow new users. This is how users join the deliberation, by adding a statement to it. 

Please fix this one, and see what else comes up. There may be things we need to discuss, I don't know.  

We may have to make some changes to the tests, or we may have to reconsider the validation checks, but in the end all the tests need to pass.

Thanks.

guodong323 commented:
OK, I will solve this problem at first.

guodong323 commented:
## **Test Results**

 **insertStatementId**

- Allows new users to join by submitting a statement.
- Ensures statements are correctly stored in `ShownStatements[0]`.
- Returns `undefined` if `discussionId` does not exist.

 **getStatementIds**

- Retrieves statement IDs correctly for users assigned statements.
- Returns `undefined` if `userId` is not part of the discussion.

 **putGroupings**

- Throws an error if a user tries to group statements they have not seen.

 **rankMostImportant**

- Prevents ranking of statements not assigned to the user.
- Successfully updates the ranking for statements that were shown.
- Ensures `rank` values are either `0` or `1`.

Here are some test results pictures
## 


![Xnip2025-02-15_15-26-58](https://github.com/user-attachments/assets/d118527c-0e24-4163-9cf7-8c9df98418e5)
![Xnip2025-02-15_15-27-19](https://github.com/user-attachments/assets/815f0980-d198-4174-ba60-65f6e4dd5601)
![Xnip2025-02-15_15-27-34](https://github.com/user-attachments/assets/b26c84d8-6d86-484c-bac8-adbb86c2cc35)
![Xnip2025-02-15_15-27-49](https://github.com/user-attachments/assets/e70a6d62-48ff-435c-ad81-26e90c4a47bb)
![Xnip2025-02-15_15-28-03](https://github.com/user-attachments/assets/863a0203-4c7f-4aae-a9fe-114fa106f05c)
![Xnip2025-02-15_15-28-33](https://github.com/user-attachments/assets/6f4f346b-81df-41f4-b272-89d0e5c1dfc6)

ddfridley commented on app/dturn/tests/dturn.js:
```diff
@@ -1,42 +1,161 @@
-  MemoryServer.stop()
-})
-
-// Tests
-test('Succeed if all options are valid.', async () => {
-  initDiscussion(DISCUSSION_ID, { group_size: 15 })
-})
-
-test('Fail if all options are not valid.', async () => {
-  const func = async () => {
-    await initDiscussion(DISCUSSION_ID, { a_nonexistent_option: 12345 })
-  }
-  expect(func()).rejects.toThrow(nonexistentOptionErrorRegex)
-})
-
-test('Fail if at least 1 option is not valid.', async () => {
-  const func = async () => {
-    await initDiscussion(DISCUSSION_ID, { group_size: 10, a_nonexistent_option: 12345 })
-  }
-  expect(func()).rejects.toThrow(nonexistentOptionErrorRegex)
-})
```
These tests seem to be missing?

guodong323 replied:
Yes, these cases have already been covered in other test cases.

ddfridley commented on app/dturn/tests/dturn.js:
```diff
@@ -1,42 +1,161 @@
+    Object.keys(Discussions).forEach(key => delete Discussions[key]);
+
+    // Initialize a test discussion with required properties
+    Discussions.testDiscussion = {
```
We should try to use initDiscussion rather than directly manipulating it.

ddfridley reviewed (changes requested):
This is looking good and I'm so glad all the dturn tests are passing.  Beside the comments inline, there's still a problem with the formatting.  I see all these semicolons at the end of lines.  We aren't using those in this project.  If you install prettier into vscode it will automatically format everything on save and we won't have to worry about it.  https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode (or find it for whatever editor you are using).

guodong323 commented on app/dturn/dturn.js:
```diff
@@ -463,40 +481,83 @@ function iteratePairs(discussionId, round, statementIds, func) {
+    return undefined;
+  }
+
+  // 确保 `statementId` 存在于 `ShownStatements`
```
Thank you, I will pay attention to this

guodong323 commented:
> This is looking good and I'm so glad all the dturn tests are passing. Beside the comments inline, there's still a problem with the formatting. I see all these semicolons at the end of lines. We aren't using those in this project. If you install prettier into vscode it will automatically format everything on save and we won't have to worry about it. https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode (or find it for whatever editor you are using).

Hi David. I already fixed these format problem. I downloaded prettier and make the format is the same with out codebase.

ddfridley commented on app/dturn/__tests__/rank-statements.js:
```diff
@@ -24,37 +16,38 @@ const OPTIONS = {
     expect(statements).toBeDefined()
     expect(statements.length).toBe(10)
 
+    console.log('Retrieved statements:', statements)
```
okay, now that this code is working isn't it time to take the console.log statements out.

ddfridley commented on package.json:
```diff
@@ -71,6 +71,7 @@
     "@babel/preset-react": "^7.16.0",
     "@svgr/cli": "^8.1.0",
     "babel-loader": "^8.2.3",
+    "prettier": "3.5.2",
```
Was prettier required to get it working in vscode?  -- are you using vs code or something else?
I don't want to have prettier as a dependency if we don't have to.  But if that was required then lets do it.  But can you uninstall prettier and see if it's working in your editor?   
Also, if it is required it should be in optionalDependencies.  You have to put it there manually in the editor.  When we push to production on heroku, devDependencies are installed and build and used to build the repo.  optionalDependencies are not installed in order to save space and build time.

ddfridley reviewed (changes requested):
This is going really well, and I see some good changes to dturn.js.  I know it's hard working in other peoples code, and this code is especially complicated. 

While all of the dturn tests pass, it turns out there are a few others when I do npm run test that don't pass.
```
npm run test app/socket-apis/__tests__/complete-round.js app/socket-apis/__tests__/post-point-groupings.js app/socket-apis/__tests__/get-points-for-round.js
```
In one case I see getStatementIds is returning [] instead of undefined, so dturn needs to be fixed for that.
For some of the others, it may be that the test needs to be fixed.  Can you look at them, and see what you can do, but if it's gets too deep then I can take a turn figuring them out.

Also, I pushed some changes, so you will need to `git pull` in this branch.  I put back some comments that were taken out, that are still valuable.   What I learned is that comments about why a piece of code is there are the most valuable.  I also changed the other throw's .to cosole.errors.  When code is being used in api calls, throws leave the api call incomplete. I know I'm the one that put the throws there. But since I was in there it was time to get rid of them.    

Have a look through this commit to see what I changed: https://github.com/EnCiv/civil-pursuit/pull/281/commits/deeb2719558a4e8519560afe1d6776455377f2eb

guodong323 commented:
> This is going really well, and I see some good changes to dturn.js. I know it's hard working in other peoples code, and this code is especially complicated.
> 
> While all of the dturn tests pass, it turns out there are a few others when I do npm run test that don't pass.
> 
> ```
> npm run test app/socket-apis/__tests__/complete-round.js app/socket-apis/__tests__/post-point-groupings.js app/socket-apis/__tests__/get-points-for-round.js
> ```
> 
> In one case I see getStatementIds is returning [] instead of undefined, so dturn needs to be fixed for that. For some of the others, it may be that the test needs to be fixed. Can you look at them, and see what you can do, but if it's gets too deep then I can take a turn figuring them out.
> 
> Also, I pushed some changes, so you will need to `git pull` in this branch. I put back some comments that were taken out, that are still valuable. What I learned is that comments about why a piece of code is there are the most valuable. I also changed the other throw's .to cosole.errors. When code is being used in api calls, throws leave the api call incomplete. I know I'm the one that put the throws there. But since I was in there it was time to get rid of them.
> 
> Have a look through this commit to see what I changed: [deeb271](https://github.com/EnCiv/civil-pursuit/commit/deeb2719558a4e8519560afe1d6776455377f2eb)

![Xnip2025-03-17_10-38-43](https://github.com/user-attachments/assets/106994c0-8801-4b49-ab66-0d15ddbf3738)
passed all tests

ddfridley commented on app/dturn/dturn.js:
```diff
@@ -204,46 +203,55 @@ function sortLargestFirst(a, b) {
-      throw new Error(
-        `getStatments unexpected number of statments ${JSON.stringify(dis.Uitems?.[userId]?.[round], null, 2)}`
-      )
+      return sIds.length > 0 ? sIds : []
```
this is a long way of saying `return sIds` - please shorten.

ddfridley commented on app/dturn/dturn.js:
```diff
@@ -336,7 +345,7 @@ async function getStatementIds(discussionId, round, userId) {
   await dis.updateUInfo(delta)
 
-  return statementIds
+  return statementIds.length > 0 ? statementIds : []
```
here is another case where it's the long way of saying `return statementIds`

ddfridley reviewed (changes requested):
- [ ] Okay, I think there was a misunderstanding about how to handle the tests failing. It is the design intent that getStatementIds returns undefined rather than [] when there are no statements to work on. If undefined is returned sometimes, and [] other times, then the code has to test for if(!ids || !ids.length) where as if it just returns undefines, then the test is just if(!ids).  Can you please change it the other way.  And FYI - there's another test.  `node app/dturn/test.js` which runs there a big loop of putting stuff in to a discussion, and then validating it.   I don't have it in jest because it takes a long time to run. 

- [ ] Please look through the changes in the [PR Files Changed](https://github.com/EnCiv/civil-pursuit/pull/281/files) and put back the comments that were taken out.  Some of them may not be appropriate anymore, but many of them are and comments are hard to get. 

- [ ] please do `git checkout origin/master package.json` and `git checkout origin/master package-lock.json`.  The changes to those files aren't relevant to this issue.
